### PR TITLE
ULTRA refactored state preprocessor

### PR DESCRIPTION
**Pull request by christianjans**

### CODE REVIEW

File: ultra/baselines/common/baseline_state_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            "images": images,
+            "low_dim_states": low_dim_states,
+            "social_vehicles": social_vehicles,
+        }
+        return out
+
+    @staticmethod
+    def _adapt_observation_for_baseline(state):
+        # Get basic information about the ego vehicle.
+        ego_position = StatePreprocessor._extract_ego_position(state)
```
Comment (adreena) [inline ultra/baselines/common/baseline_state_preprocessor.py]: variables or methods starting with _ are usually to denote private variables/modules I suggest changing them to without _

File: ultra/baselines/common/baseline_state_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            num_lookahead=observation_num_lookahead,
+            goal_path=state["goal_path"],
+            ego_position=state["ego_position"],
+            ego_heading=state["heading"],
+        )
+        state["waypoints_lookahead"] = np.hstack(lookahead_waypoints)
+
+        # Normalize the low-dimensional states and concatenate them.
+        normalized = [
+            BaselineStatePreprocessor._normalize(key, state[key])
```
Comment (adreena) [inline ultra/baselines/common/baseline_state_preprocessor.py]: self._normalize()?

Comment (adreena) [reply ultra/baselines/common/baseline_state_preprocessor.py]: could you explain why some methods are static?

File: ultra/baselines/common/baseline_state_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        # Identify the path the ego is following.
+        ego_goal_path = StatePreprocessor._extract_ego_path(
+            goal=ego_goal,
+            waypoints=ego_waypoints,
+            start=ego_start,
+        )
+
+        # Get the closest waypoint to the ego.
+        ego_closest_waypoint, _ = StatePreprocessor._extract_closest_waypoint(
```
Comment (adreena) [inline ultra/baselines/common/baseline_state_preprocessor.py]: same comment, could it be self.extract_closest_waypoint() since it's inheriting from StatePreprocessor?

File: ultra/baselines/common/baseline_state_preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            image = image.unsqueeze(0) if unsqueeze else image
+            image = image.to(device) if device else image
+            image = normalize_im(image) if normalize else image
+            images[k] = image
+
+        # Set the action in the basic state as the previous action.
+        state["action"] = prev_action
+
+        # Obtain the next waypoints.
+        _, lookahead_waypoints = get_closest_waypoint(
```
Comment (adreena) [inline ultra/baselines/common/baseline_state_preprocessor.py]: Could you move the get_closest_waypoint to the StatePreprocessor as well? to follow the same design with extra_modules()? I think it would be cleaner that way

File: ultra/baselines/common/baseline_state_preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        ]
+        low_dim_states = torch.cat(
+            [torch.from_numpy(e).float() for e in low_dim_states], dim=-1
+        )
+        low_dim_states = low_dim_states.unsqueeze(0) if unsqueeze else low_dim_states
+        low_dim_states = low_dim_states.to(device) if device else low_dim_states
+
+        # Apply the social vehicle encoder (if applicable).
+        state["social_vehicles"] = (
+            get_social_vehicles(
```
Comment (adreena) [inline ultra/baselines/common/baseline_state_preprocessor.py]: Could get_social_vehicles also go inside StatePreprocessor?

File: ultra/baselines/common/custom_state_preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    """An example of how one would implement their own state preprocessor."""
+    def __init__(self, speed_normalization):
+        self._speed_normalization = speed_normalization
+
+    def _preprocess_state(self, state):
+        speed = StatePreprocessor._extract_ego_speed(state)
+        normalized_speed = self._speed_normalization(speed)
+        position = StatePreprocessor._extract_ego_position(state)
+        steering = StatePreprocessor._extract_ego_steering(state)
+        heading = StatePreprocessor._extract_ego_heading(state)
```
Comment (adreena) [inline ultra/baselines/common/custom_state_preprocessor.py]: if you move the social_vehicle extraction to StatePreprocessor you could also include it as your example of CustomStatePreprocessor

File: ultra/baselines/common/state_preprocessor.py [hunk 7]
```diff
@@ -19,190 +19,145 @@
...
-            observation_num_lookahead=observation_num_lookahead,
-            social_capacity=social_capacity,
-            normalize=normalize,
-            unsqueeze=unsqueeze,
-            device=device,
-            convert_action_func=self.convert_action_func,
-            social_vehicle_config=social_vehicle_config,
-            prev_action=prev_action,
-            draw=draw,
+    def __call__(self, state, *args, **kwargs):
```
Comment (adreena) [inline ultra/baselines/common/state_preprocessor.py]: 👍🏻

File: ultra/baselines/common/state_preprocessor.py [hunk 8]
```diff
@@ -19,190 +19,145 @@
...
+            Parameters:
+                state (Observation): The raw environment observation.
+
+            Returns:
+                (float): The ego vehicle's speed in meters per second.
+        """
+        return state.ego_vehicle_state.speed
+
+    @staticmethod
+    def _extract_ego_steering(state: Observation):
```
Comment (adreena) [inline ultra/baselines/common/state_preprocessor.py]: nice I like type scripting here it would let the user know they should pass raw observation object 👍🏻

File: ultra/baselines/ddpg/ddpg/policy.py [hunk 9]
```diff
@@ -93,16 +93,13 @@ def __init__(
...
+        self.state_preprocessor = BaselineStatePreprocessor(
             policy_params["social_vehicles"],
             policy_params["observation_num_lookahead"],
             self.action_size,
         )
-        self.state_preprocessor = StatePreprocessor(
-            preprocess_state, to_2d_action, self.state_description
-        )
+
+        self.social_vehicle_encoder = self.social_vehicle_config["encoder"]
```
Comment (adreena) [inline ultra/baselines/ddpg/ddpg/policy.py:102]: Could the encoder be moved to StatePreprocessor? as state_preprocessor.social_vehicle_encoder()?
we need to implement it in a way to allow CustomeStatePreprocessors override the encoder too

**Comment by adreena** [comment]
Good job on refactoring 🥇  I left a couple of questions and suggestions

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 1 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Okay sure, sounds good.

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 4 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Yeah okay definitely, will do this.

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 5 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Oh for sure yeah, this would be super helpful. I will look into doing this.

File: ultra/baselines/common/custom_state_preprocessor.py (hunk 6 above)
Comment (christianjans) [reply ultra/baselines/common/custom_state_preprocessor.py]: Perfect yes, I will move social_vehicle_extraction.py capability to the StatePreprocessor.

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 2 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Oh yeah for sure, the static methods in BaselineStatePreprocessor and StatePreprocessor are static because they don't require any instance attributes to perform the method. What are your thoughts on this?

Additionally, yes I agree it should be self._normalize(...) and also self.extract_closest_waypoint(...) in the next comment. Then should other calls to static methods like in _adapt_observation_for_baseline and _normalize be replaced with self.<method_name>? I guess this would mean that _adapt_observation_for_baseline and _normalize would no longer be static methods.

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 3 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Ahh right, yes definitely; see the previous response for more detail.

File: ultra/baselines/ddpg/ddpg/policy.py (hunk 9 above)
Comment (christianjans) [reply ultra/baselines/ddpg/ddpg/policy.py:102]: Yeah, I was wondering about that too. Do you mean just the metadata about the encoder like social_vehicle_config be a part of the StatePreprocessor? Or the actual encoder neural network as well?

If the actual encoder network is part of the state preprocessing, would we have to worry about the encoder network using PyTorch and the network used for the agent using another library like TensorFlow? Or perhaps this would be a problem for the ULTRA user... they would either have to make their own encoder or make their agent use a PyTorch network.

Additionally, if the actual encoder network was used in the state preprocessing to encode the social vehicles before sending it back to the agent, the agent would lose control over this encoder network. By the phrase 'lose control', I mean the agent's policy class could not set the state or properties of the encoder network. For example in the DQN, when it learns, this call to calculate `qs_network_target` without gradient calculations would have no effect on the encoder network. That is, the encoder network would still calculate the gradients since it would be a separate network from the agent's network:
https://github.com/christianjans/SMARTS/blob/505b10939e92f504921e18da3f26f54ca08d1f08/ultra/baselines/dqn/dqn/policy.py#L344-L346
Similarly, these calls to train() and eval() wouldn't set the encoder network to train() or eval() mode, it would just set the agent's network to these modes:
https://github.com/christianjans/SMARTS/blob/505b10939e92f504921e18da3f26f54ca08d1f08/ultra/baselines/dqn/dqn/policy.py#L349
https://github.com/christianjans/SMARTS/blob/505b10939e92f504921e18da3f26f54ca08d1f08/ultra/baselines/dqn/dqn/policy.py#L349

Let me know if I can clarify anything or if something I said isn't correct.

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 2 above)
Comment (adreena) [reply ultra/baselines/common/baseline_state_preprocessor.py]: the normalize method is just used for baselines, other people may need to define their own parameters for normalization I think it doesn't need to be static, is there anywhere in the code that calls the normalization? if not it's safe to keep it a method inside baseline_state_preprocessor. The same idea for adapt module, if others decide to use baseline they're gonna use the entire preprocessor not just one of the methods (at least I'm not foreseeing it)

File: ultra/baselines/ddpg/ddpg/policy.py (hunk 9 above)
Comment (adreena) [reply ultra/baselines/ddpg/ddpg/policy.py:102]: the encoders are processed to work with pytorch, and the encoding is not happening in the statepreprocessing, statepreprocessor only extracts the vehicle information to feed the networks... lets not touch it for now, make another task to brainstorm the encoders later

File: ultra/baselines/common/baseline_state_preprocessor.py (hunk 2 above)
Comment (christianjans) [reply ultra/baselines/common/baseline_state_preprocessor.py]: Ahh, I see what you mean. Sure, I will make them both instance methods.

File: ultra/baselines/ddpg/ddpg/policy.py (hunk 9 above)
Comment (christianjans) [reply ultra/baselines/ddpg/ddpg/policy.py:102]: Will do 👍
